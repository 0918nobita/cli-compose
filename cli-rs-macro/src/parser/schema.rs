use std::fmt;

use syn::{
    parse::{Parse, ParseStream},
    Ident, Token,
};

use super::{field_schema::FieldSchema, modifiers::Modifiers};
use crate::parser::{arg_kind::ArgKind, modifier::Modifier};

pub struct Schema {
    pub kind: ArgKind,
    pub modifiers: Modifiers,
    pub data: SchemaData,
}

impl fmt::Debug for Schema {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "Schema({:?}, {:?}, {:?})",
            self.kind, self.modifiers, self.data
        )
    }
}

#[derive(Debug)]
pub enum SchemaData {
    Single(FieldSchema),
    Multiple(Vec<FieldSchema>),
}

impl Parse for Schema {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let ident = input.parse::<Ident>()?;

        let kind = ArgKind::try_from(ident)?;

        let modifiers = if input.peek(syn::token::Paren) {
            let parenthesized;
            syn::parenthesized!(parenthesized in input);

            let modifiers = parenthesized
                .parse_terminated::<Modifier, Token![,]>(Modifier::parse)?
                .into_iter()
                .collect::<Vec<_>>();

            match Modifiers::try_from_slice(modifiers.as_slice()) {
                Ok(modifiers) => modifiers,
                Err(modifier) => {
                    return Err(syn::Error::new_spanned(modifier.name, "Duplicate modifier"))
                }
            }
        } else {
            Modifiers::default()
        };

        let data = if input.peek(syn::token::Brace) {
            let braced;
            syn::braced!(braced in input);

            let field_schemas = braced
                .parse_terminated::<FieldSchema, Token![,]>(FieldSchema::parse)?
                .into_iter()
                .collect::<Vec<_>>();

            SchemaData::Multiple(field_schemas)
        } else {
            let field_schema = input.parse::<FieldSchema>()?;
            SchemaData::Single(field_schema)
        };

        Ok(Self {
            kind,
            modifiers,
            data,
        })
    }
}
