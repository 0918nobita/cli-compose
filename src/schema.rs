use derive_more::Display;

pub use cli_compose_macro::{ArgOpt, Cli, FromKebabStr, MultiSelect, Opt, PosArg, SingleSelect};

#[derive(Display)]
pub enum Flag {
    #[display(fmt = "--{}", _0)]
    LongOnly(String),

    #[display(fmt = "-{}", _0)]
    ShortOnly(char),

    #[display(fmt = "--{}, -{}", _0, _1)]
    BothLongAndShort(String, char),
}

pub trait AsPosArg: Sized {
    fn name() -> String;

    fn description() -> String;

    fn parse(s: &str) -> Option<Self>;
}

pub trait AsArgOpt: Sized {
    fn flag() -> Flag;

    fn description() -> String;

    fn parse(s: &str) -> Option<Self>;
}

pub trait AsOpt {
    fn flag() -> Flag;

    fn description() -> String;
}

pub trait AsSingleSelect {
    fn name() -> String;

    fn description() -> String;
}

pub trait AsMultiSelect {
    fn name() -> String;

    fn description() -> String;
}
