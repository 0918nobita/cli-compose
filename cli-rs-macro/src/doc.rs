use syn::{Attribute, Lit, Meta};

fn try_get_single_line_doc(attr: &Attribute) -> Option<String> {
    let meta = attr.parse_meta().ok()?;

    let lit_str = match meta {
        Meta::NameValue(syn::MetaNameValue {
            path,
            lit: Lit::Str(lit_str),
            ..
        }) if path.is_ident("doc") => lit_str,
        _ => return None,
    };

    Some(lit_str.value().trim_start().to_owned())
}

pub fn extract_doc(attrs: &[Attribute]) -> String {
    attrs
        .iter()
        .filter_map(try_get_single_line_doc)
        .collect::<Vec<_>>()
        .join("\n")
}
