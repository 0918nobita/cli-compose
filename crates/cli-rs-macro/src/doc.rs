use syn::Attribute;

fn try_get_single_line_doc(attr: Attribute) -> Option<String> {
    let meta = attr.parse_meta().ok()?;
    match meta {
        syn::Meta::NameValue(syn::MetaNameValue {
            path,
            lit: syn::Lit::Str(lit_str),
            ..
        }) if path.is_ident("doc") => Some(lit_str.value().trim_start().to_owned()),
        _ => None,
    }
}

pub fn extract_doc(attrs: impl Iterator<Item = Attribute>) -> String {
    attrs
        .filter_map(try_get_single_line_doc)
        .collect::<Vec<_>>()
        .join("\n")
}
