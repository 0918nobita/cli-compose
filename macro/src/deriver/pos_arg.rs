use bae::FromAttributes;
use convert_case::{Case, Casing};
use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use syn::Data;

use crate::doc::extract_doc;

#[derive(FromAttributes)]
struct PosArg {
    name: Option<syn::LitStr>,

    #[allow(dead_code)]
    use_default: Option<()>,
}

pub fn derive_pos_arg(input: TokenStream) -> syn::Result<TokenStream> {
    let input = syn::parse2::<syn::DeriveInput>(input)?;

    let attr = PosArg::try_from_attributes(&input.attrs)?;

    let doc = extract_doc(&input.attrs);

    let ty_name = &input.ident;
    let ty_name_str = ty_name.to_string();
    let ty_name_kebab_case = attr.and_then(|attr| attr.name).map_or_else(
        || ty_name_str.to_case(Case::Kebab),
        |lit_str| lit_str.value(),
    );

    let methods = match &input.data {
        Data::Enum(_) => {
            quote! {
                fn parse(s: &str) -> Option<Self> {
                    <#ty_name as std::str::FromStr>::from_str(s).ok()
                }

                fn result() -> cli_compose::schema::forwarded::syn::Type {
                    cli_compose::schema::forwarded::syn::parse_str(#ty_name_str).unwrap()
                }
            }
        }

        Data::Struct(data_struct) => match data_struct.fields.iter().collect::<Vec<_>>()[..] {
            [field] => {
                let ty = &field.ty;
                let ty_name = ty.into_token_stream().to_string();
                let parse_method = if let Some(ident) = &field.ident {
                    quote! {
                        fn parse(s: &str) -> Option<Self> {
                            <#ty as std::str::FromStr>::from_str(s).ok().map(|v| Self { #ident: v })
                        }
                    }
                } else {
                    quote! {
                        fn parse(s: &str) -> Option<Self> {
                            <#ty as std::str::FromStr>::from_str(s).ok().map(Self)
                        }
                    }
                };
                quote! {
                    #parse_method

                    fn result() -> cli_compose::schema::forwarded::syn::Type {
                        cli_compose::schema::forwarded::syn::parse_str(#ty_name).unwrap()
                    }
                }
            }
            [] => return Err(syn::Error::new_spanned(input.ident, "missing field")),
            _ => {
                return Err(syn::Error::new_spanned(
                    data_struct.struct_token,
                    "multiple fields are not allowed",
                ))
            }
        },

        Data::Union(data_union) => {
            return Err(syn::Error::new_spanned(
                data_union.union_token,
                "unions are not allowed",
            ));
        }
    };

    let sharp = syn::Token![#](proc_macro2::Span::call_site());

    Ok(quote::quote! {
        impl cli_compose::schema::AsMember for #ty_name {
            fn handle(mut builder: cli_compose::schema::CliBuilder) -> cli_compose::schema::CliBuilder {
                use cli_compose::schema::{forwarded::{syn, quote}, AsPosArg};

                let name = <#ty_name as AsPosArg>::name();

                let res_ty =
                    <syn::Type as quote::ToTokens>::into_token_stream(
                        <#ty_name as AsPosArg>::result()
                    ).to_string();

                builder.ops.extend(quote::quote! {
                    println!("PosArg {} ({})", #sharp name, #sharp res_ty);
                });

                builder
            }
        }

        impl cli_compose::schema::AsPosArg for #ty_name {
            fn name() -> String {
                #ty_name_kebab_case.to_owned()
            }

            fn description() -> String {
                #doc.to_owned()
            }

            #methods
        }
    })
}

#[cfg(test)]
mod tests {
    use quote::quote;

    fn test_derive_pos_arg(input: proc_macro2::TokenStream) -> anyhow::Result<String> {
        let tokens = super::derive_pos_arg(input)?;

        crate::pretty_print::pretty_print_rust_code(tokens)
    }

    #[test]
    fn empty() {
        insta::assert_debug_snapshot!(test_derive_pos_arg(quote! {}));
    }

    #[test]
    fn struct_without_field() {
        insta::assert_debug_snapshot!(test_derive_pos_arg(quote! {
            struct Foo;
        }));
    }

    #[test]
    fn struct_with_single_field() {
        insta::assert_display_snapshot!(test_derive_pos_arg(quote! {
            struct Foo(String);
        })
        .unwrap());
    }

    #[test]
    fn struct_with_multiple_fields() {
        insta::assert_debug_snapshot!(test_derive_pos_arg(quote! {
            struct Foo(String, i32);
        }));
    }
}
