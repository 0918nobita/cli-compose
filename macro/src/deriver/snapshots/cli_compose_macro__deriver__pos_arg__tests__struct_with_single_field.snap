---
source: macro/src/deriver/pos_arg.rs
assertion_line: 146
expression: "test_pos_arg_deriver(quote! { struct Foo(String) ; }).unwrap()"
---
impl cli_compose::schema::AsMember for Foo {
    fn handle(mut builder: cli_compose::schema::CliBuilder) -> cli_compose::schema::CliBuilder {
        use cli_compose::schema::{
            forwarded::{quote, syn},
            AsPosArg,
        };
        let name = <Foo as AsPosArg>::name();
        let res_ty = <syn::Type as quote::ToTokens>::into_token_stream(<Foo as AsPosArg>::result())
            .to_string();
        builder
            .ops
            .extend(quote::quote! { println ! ("PosArg {} ({})" , # name , # res_ty) ; });
        builder
    }
}
impl cli_compose::schema::AsPosArg for Foo {
    fn name() -> String {
        "foo".to_owned()
    }
    fn description() -> String {
        "".to_owned()
    }
    fn parse(s: &str) -> Option<Self> {
        <String as std::str::FromStr>::from_str(s).ok().map(Self)
    }
    fn result() -> cli_compose::schema::forwarded::syn::Type {
        cli_compose::schema::forwarded::syn::parse_str("String").unwrap()
    }
}

